### ISSUE TITLE
Fluentmark preview is not using the same markdown as GitLab

### ISSUE THREAD

**Issue by ileslie** [grosenberg/Fluentmark#70]
I have Fluentmark setup in Eclipse and can happily edit markdown documents. I have a problem with the PlanUML integration though.

When I create markdown that renders my PlantUML diagram correctly on the GitLab Wiki it requires that the mark down looks like this (note obviously without the extra space in the back tick trio at the end):
```
## Audience

This document is a test for Fluentmark plantuml rendering.

```plantuml
@startuml
	Bar -> Foo : hello
@enduml
	
@startuml
	Ernie -> Bert : heyman
@enduml
`` `
```

But that doesn't do anything helpful in the Fulentmark preview. The preview just renders the instructions as if it ware rendering pretty printed java or c code.
![UMLNotRendered](https://user-images.githubusercontent.com/4226105/231793047-ba4b1576-c394-4121-b4d7-14fc783921c7.png)

I can change the markdown for Fluentmark's preview to work by doing this;
```
## Audience

This document is a test for Fluentmark plantuml rendering.

@startuml
	Bar -> Foo : hello
@enduml
	
@startuml
	Ernie -> Bert : heyman
@enduml
```
My preview looks fine:
![UMLRendered](https://user-images.githubusercontent.com/4226105/231793094-7b05b0d3-9c0a-4dca-bdd6-ed8306f32275.png)

But on GitLab it does not render the UML diagram. What do I have setup incorrectly or is this a bug?

Versions:
Eclipse IDE for C/C++ Developers Version: 2023-03 (4.27.0)

FluentMark Editor	1.6.4.20210420-2328	net.certiv.fluent.dt.feature.feature.group	Certiv Analytics

pandoc.exe 3.1.2 Features: +server +lua Scripting engine: Lua 5.4

dot - graphviz version 8.0.2 (20230410.1723)

**Comment by grosenberg** [comment]
You are doing everything correct. Just that PlantUML support in markdown is completely non-standard.

Fluentmark supports PlantUML (along with LaTex and Dot) like HTML as first-class markdown entities requiring no separate markup. As you noticed, marking PlantUML source in a source code block (```) yields a source code presentation just as you should expect, i.e., just like how every other source code block works.

Hard to justify changing Fluentmark's behavior absent some consensus on handling.

**Comment by ileslie** [comment]
Interesting but how am I supposed to use your editor? In the end my documents need to be rendered on line in GitLab. If I cannot see what my document looks like in your editor then I cannot use your editor to edit my documents. After all the goal is not to render the markdown in an editor the goal is to render documents on a server.

I am not sure how anyone can move forward who's using GitLab to hold their projects. Aren't there other people whose renderer does the same thing as GibLab? I'm not arguing about who's right in principal - I honestly don't care - I'm wondering how people use your editor if their renderer doesn't work with that syntax.

**Comment by grosenberg** [comment]
Not really too familiar with GitLab & was not aware of any repository provider, including GitLab, that offered PlantUML rendering.

To be certain, the GitLab markdown is not portable.  Given the lack of any standard, don't know how to directly address the issue.

The only present option is that Fluentmark does support both HTML and PDF productions that embed the rendered PlantUML code as SVG images.

I will leave this open for now as an enhancement request

**Comment by geodab** [comment]
For what it is worth, I just happened to be looking at this same issue and am happy to see this project seeing some activity. I have enjoyed using this plugin in my workflow.

As a workaround, to use Fluentmark Live View while also enabling a published view in GitLab I started applying a simple awk filter to wrap the bare plantuml in a secondary file:
```awk
/@startuml/ { print "```plantuml" }
{ print }
/@enduml/ { print "```" }
```
This filtered version of the file is also what I have used to run pandoc directly to produce a pdf or html outside of eclipse.

Digging deeper, the CommonMark spec for fenced code blocks indeed leaves the interpretation of an [info string](https://spec.commonmark.org/0.30/#fenced-code-blocks) undefined.  However, it looks like using pandoc with the [pandoc-plantuml](https://github.com/timofurrer/pandoc-plantuml-filter/tree/v0.1.1) filter matches the behavior seen in GitLab, so perhaps that is what GitLab is using?  That filter processes code blocks with class `plantuml` taken by pandoc from the info string, which seems like a reasonable approach even if not exactly portable since there is no standard.  For another data point, while GitHub does not support plantuml directly it does support similar code blocks using `mermaid`, and GitLab supports that too, so the syntax appears to be fairly widespread.

As an enhancement, would it be possible to provide an option to have the Live View match the pandoc filtered output when it is chosen as the Markdown export converter in the preferences?  Does Fluentmark look at the info string for the code block so that it could process it like it does with a bare plantuml block for that case?

**Comment by grosenberg** [comment]
I can modify Fluentmark to follow the consensus. 

>  while GitHub does not support plantuml directly it does support similar code blocks using mermaid

Did not know that -- not documented in the formal [GitHub Flavored Markdown Spec](https://github.github.com/gfm/), but is described in a somewhat obscure [blog post](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/)

Anyone have anything that says the GitHub blog post syntax is authoritative to GitHub or CommonMark?

Anyone have any other examples of how/where diagram embedding is implemented?

**Comment by geodab** [comment]
I found this at GitHub referring to creating diagrams as "advanced formatting":  [Creating Mermaid diagrams](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams)

And a similar reference at GitLab: [Diagrams and flowcharts](https://docs.gitlab.com/ee/user/markdown.html#diagrams-and-flowcharts)

It appears that GitLab has been pushing the markdown support a bit further with the [GLFM](https://docs.gitlab.com/ee/development/gitlab_flavored_markdown/specification_guide/) spec by building on the GFM spec.  As part of that effort they have also noted that it seems that the GFM spec is out of date: https://github.com/github/cmark-gfm/issues/288

You may also be interested in looking at https://kroki.io/ which is available in GitLab to provide a wider range of diagramming support.  There is a [Markdown Kroki](https://marketplace.visualstudio.com/items?itemName=pomdtr.markdown-kroki) plugin for VSCode that uses a similar code fence syntax for the many formats it supports.

**Comment by grosenberg** [comment]
Released version 1.8 now supports both diagram variants: native and codeblock.

**Comment by geodab** [comment]
Thanks for adding this, it is much appreciated.  It took some time for me to update to 1.8.0 but I can now confirm that the markdown code fence syntax is working as expected in both the Live View and in GitLab.
